### PR TITLE
Support tuple types in the human readable abi

### LINKED ISSUE #208: Support tuples in humanReadableAbi

### PR DESCRIPTION
**Pull request by sz-piotr**
Closes #208.

### CODE REVIEW

File: waffle-compiler/src/getHumanReadableAbi.ts [hunk 1]
```diff
@@ -27,31 +27,40 @@ function encodeEvent(entry: any) {
...
+    const components = value.components.map(encodeSolidityValue).join(', ');
+    result = result.replace('tuple', `tuple(${components})`);
+  }
+  if (value.indexed) {
+    result += ' indexed';
+  }
+  if (value.name) {
+    result += ` ${value.name}`;
+  }
+  console.log(result);
```
Comment (marekkirejczyk) [inline waffle-compiler/src/getHumanReadableAbi.ts:64]: meh

File: waffle-compiler/src/getHumanReadableAbi.ts [hunk 2]
```diff
@@ -27,31 +27,40 @@ function encodeEvent(entry: any) {
...
-      }
-      return output.type;
-    })
-    .join(', ');
+  const returns = outputs.map(encodeSolidityValue).join(', ');
   return ` returns(${returns})`;
 }
+
+function encodeSolidityValue(value: SolidityValue) {
+  let result = value.type;
```
Comment (marekkirejczyk) [inline waffle-compiler/src/getHumanReadableAbi.ts:53]: `let {type} = value;`

File: waffle-compiler/src/getHumanReadableAbi.ts [hunk 3]
```diff
@@ -27,31 +27,40 @@ function encodeEvent(entry: any) {
...
+  if (result.includes('tuple') && value.components) {
```
Comment (marekkirejczyk) [inline waffle-compiler/src/getHumanReadableAbi.ts:54]: includes? not equals?

Comment (sz-piotr) [reply waffle-compiler/src/getHumanReadableAbi.ts:54]: Has to be includes, because of stuff like `tuple[]`

**Comment by marekkirejczyk** [comment]
Released in [2.4.1](https://github.com/EthWorks/Waffle/releases/tag/2.4.1).

**Pull request merged**